embassy-rp/src/pio_programs/clock_divider.rs

```diff
@@ -14,12 +13,62 @@ use crate::clocks::clk_sys_freq;
 /// # Returns
 ///
 /// A fixed-point divider value suitable for use in a PIO state machine configuration
-#[inline]
 pub fn calculate_pio_clock_divider(target_hz: u32) -> fixed::FixedU32<U8> {
+    calculate_pio_clock_divider_value(clk_sys_freq(), target_hz)
+}
+
+/// Calculate a PIO clock divider value based on the desired target frequency.
+///
+/// # Arguments
+///
+/// * `sys_hz` - The system clock frequency in Hz
+/// * `target_hz` - The desired PIO clock frequency in Hz
+///
+/// # Returns
+///
+/// A fixed-point divider value suitable for use in a PIO state machine configuration
+pub const fn calculate_pio_clock_divider_value(sys_hz: u32, target_hz: u32) -> fixed::FixedU32<U8> {
     // Requires a non-zero frequency
-    assert!(target_hz > 0, "PIO clock frequency cannot be zero");
+    core::assert!(target_hz > 0);
```

Comment:
use `assert!` so panic messages are properly logged with defmt if enabled.

Reply:
Thanks for having a look!

Unfortunately, `assert!` will not compile in a `const fn` if the `defmt` feature is enabled. This is what I meant in a previous comment:
> I hit compile errors with defmt::assert! in the const fn. Resolved by using core::assert without a string literal instead; not sure if that's the right approach. I could add an error enum, or split this into three functions (runtime-with-defmt, const-without-defmt, and math-inner), but maybe that would be overkill?
> 
> Or I could just drop the public const fn idea entirely, and this will always be calculated at runtime.